website/blog/2023-09-06-optimize-try-cast.mdx

```diff
@@ -0,0 +1,114 @@
+---
+slug: optimize-try_cast.
+title: "Learnings from optimizing try_cast"
+authors: [lsakka]
+tags: [tech-blog, performance]
+---
+
+One of the queries shadowed internally at Meta was much slower in Velox compared to presto(2 days vs 4.5h). Initial investigation identified that the overhead is related to casting empty strings inside a try_cast.
+
+In this blogpost I summarize my learnings from investigating and optimizing try_cast.
+
+--------------------------------------------------------------------------------------------
+
+### Start and end results
+
+Initial benchmark:
+```
+try_cast(empty_string_col as int)                     4.88s   204.83m
+try_cast(valid_string_col as int)                    2.15ms    465.22
+```
+
+The difference between casting a valid and invalid input is huge (>1000X), although ideally casting an invalid string shall be
+just setting a null and should not be that expensive.
+
+End results:
+```
+try_cast(empty_string_col as int)                    1.24ms    808.79
+try_cast(valid_string_col as int)                    2.15ms    465.22
+```
+--------------------------------------------------------------------------------------------
+
+### Sources of regressions
+The investigation revealed several factors that contributed to the huge gap, summarized in the points below in addition to
+their approximate significance.
+
+
+**Error logs overhead.**
+
+Whenever a VeloxUserError is thrown an error log used to be generated, however those errors are expected to, (1) either get converted to null if is
```

Comment:
Is the error logging disabled only for try_cast usage ? That's alright.

For regular cast, so long as the failures get propagated to event listener and available from those logs, that should suffice for any further tracing.

Reply:
we disabled the error logging for velox_user_errors for the reason specified in the note.
 https://github.com/facebookincubator/velox/pull/5610